src/program/types/ForOfStatement.js

```diff
@@ -52,19 +52,18 @@ export default class ForOfStatement extends LoopStatement {
 		code.prependRight(this.right.start, `var ${key} = 0, ${list} = `);
 		code.appendLeft(this.right.end, `; ${key} < ${list}.length; ${key} += 1`);
 
-		// destructuring. TODO non declaration destructuring
-		const declarator =
-			this.left.type === 'VariableDeclaration' && this.left.declarations[0];
-		if (declarator && declarator.id.type !== 'Identifier') {
+		const isDeclaration = this.left.type === 'VariableDeclaration';
+		const declarator = isDeclaration && this.left.declarations[0];
+		if ((declarator && declarator.id.type !== 'Identifier') || this.left.type === 'ObjectPattern') {
```

Comment:
This doesn't work for array patterns, right? Maybe better like this:
```js
const isDeclaration // …
const maybeDestructuring = isDeclaration ? this.left.declarations[0].id : this.left
if (maybeDestructuring.type !== 'Identifier') {
  destructure(
    // …
    maybeDestructuring,
    // …
  );
}
```

Reply:
Much cleaner. I'd forgotten about other desctructuring types.